[plugin] Chrony NTP service check

bbczeuz opened the pull request:
Hi,
I wrote a new plugin that collects statistics from chronyd. Chronyd is an implementation of NTP. It's the default NTP service on RHEL7, but packages for various other platforms exist (like Debian).

Tested on RHEL7 (x86_64), FreeBSD 10.2 (x86_64) and Debian 8.2.

rubenk commented:
Can you please rebase on current master?

bbczeuz commented:
@rubenk Rebased to my fork of collectd/master which I rebased to the upstream... The operation itself didn't work very smooth (had to skip many patches, added/removed files from other branches,...). Can you use it or do you prefer a plain diff? Build works on all platforms at jenkins :)

rubenk commented:
@bbczeuz first of all thanks for your contribution! These are a lot of changes which makes it somewhat hard to review. Not all of them seem to be needed and introduce unneccesary churn, for instance 9830d2a is a merge commit which just changes the file mode of configure.ac, and 4f36893 changes the collectd-nagios code which is probably an accident? I also see references to a tcpsock plugin and a wxt plugin?

I'd prefer if you squash your commits down to one, or a few commits that introduce the changes one by one.

bbczeuz commented:
Ok, I start liking ```git rebase -i``` ==> History cleaned. Do you want it even more compressed? I messed my branches in the past (I'm working on a wxt plugin (wxt520 weather sensor) and an extension to collectd-nagios to support TCP (#1053, open since 9 months...))

- I use ```/dev/urandom``` to avoid seeding the RNG via time(NULL)
- I use ```__attribute__((packed))``` to enforce/_hint_ the compiler to pack structs

rubenk commented:
Much better, thanks!

I'll leave the in-depth review to others more equipped to do that, but I'll leave some comments inline.

bbczeuz commented:
@rubenk: Thanks for the review. I implemented most of your comments. Except:
- "return (1)": I used that form it in the past (a now squashed commit), but removed the braces again as they look ugly to me.
- Indentation: I used vim's "gg=G" to enforce it's default which is tabbed. If there /is/ a strong opposition against tabs, I can change it to spaces.

rubenk commented:
Thanks for the update.

As for the coding style, we do have an existing coding style for a reason, so yes, there's s strong preference for new code to adhere to that style, as I think goes for most projects.

bbczeuz commented:
@rubenk, @mfournier 
- Rebased to current head.
- Fixed intendation to two spaces (as style guide suggests)

rubenk commented on src/Makefile.am:
```diff
@@ -216,6 +216,13 @@ cgroups_la_LDFLAGS = $(PLUGIN_LDFLAGS)
+pkglib_LTLIBRARIES += chrony.la
+chrony_la_SOURCES = chrony.c
+chrony_la_LDFLAGS = $(PLUGIN_LDFLAGS)
+chrony_la_LIBADD =
```
there's no need to set this.

bbczeuz replied:
You mean the empty LIBADD? There are several other plugins defining an empty LIBADD, too.

rubenk replied:
Yes. The other empty LIBADDs are extended later on, which is not the case here.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+ **********************************************************************
+ * Copyright (C) Claudius M Zingerli, ZSeng, 2015-2016 
+ *
+ * Internas roughly based on the ntpd plugin
```
Internals

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+ * 
+ **********************************************************************
+*/
+/* Formated using 'GNU indent -bli0 -ce -ppi 2' and some manual alignment
```
There's no use for this.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ * 
+ **********************************************************************
```
Please remove all these starts and just continue the comment. Other plugins don't do this either.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+
+#include "config.h"
+
+#if HAVE_SYS_TYPES_H
```
this is already included by collectd.h below

bbczeuz replied:
Sorry, I don't understand what you mean. HAVE_SYS_TYPES_H is usually defined in collectd.h, but there are no includes in config.h.

rubenk replied:
You already include collectd.h below, which includes `<sys/types.h>`

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+#  include <sys/types.h>        /* getaddrinfo */
+#endif
+#if HAVE_SYS_SOCKET_H
+#  include <sys/socket.h>
```
same here.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+};
+
+static int g_config_keys_num = STATIC_ARRAY_SIZE(g_config_keys);
+static int g_is_connected = 0;
```
no need to zero-initialize static variables.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+static int g_is_connected = 0;
+static int g_chrony_socket = -1;
+static time_t g_chrony_timeout = -1;
+static char *g_chrony_host = NULL;
```
same here.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+static char *g_chrony_port = NULL;
+static char *g_plugin_instance = NULL;
+static uint32_t g_chrony_rand = 1;
+static uint32_t g_chrony_seq_is_initialized = 0;
```
and here.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+} eDaemonReplies;
+
+#define ATTRIB_PACKED __attribute__((packed))
+typedef struct ATTRIB_PACKED
```
I think it's more clear if you used __attribute__((packed)) directly.

bbczeuz replied:
I use this at 13 places in the code. As this is a gcc/clang extension (=not c standard AFAIK), I prefer to define this at one place so it's easier to adapt to other compilers. Shall I test for GCC and issue a warning? Might break compilation due to ```-Werror```. Just stay quiet?
```


#if defined(__GNUC__) or defined(__clang__)
#  define ATTRIB_PACKED __attribute__((packed))
#else
#  warning Not defining packed attribute (unknown compiler)
#  define ATTRIB_PACKED 
#endif
```

rubenk replied:
Fair enough. Ideally we detect if __attribute__((packed)) is supported by the compiler (Sun Studio has it too iirc). We should do that for the other attributes we use in the codebase to, but let's leave that for another day :)

It's probably good to break compilation, you might even turn the warning into an error, so we get reports for other compilers.

One last thing, clang also defines __GNUC__ so there's no need to test for that seperately.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+  case IPADDR_INET6:
+    ip6 = addr->addr.ip6;
+
+#ifdef FEAT_IPV6
```
where does this define come from?

bbczeuz replied:
It was in the original code from chronyd. There it was statically defined in ```configure```. I removed the ```#if```, assuming inet_ntop is always defined (Didn't find any feature test macro). I'll push an updated code today evening after some testing on my systems.

rubenk replied:
I think you're reasonably safe assuming inet_ntop exists on all the platforms we build on.
If not, we can always add a check to configure.ac later.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+static int
+chrony_set_timeout()
+{
+  /*Set the socket's  timeout to g_chrony_timeout; a value of 0 signals infinite timeout */
```
Please add a space after the /*

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+  if (g_chrony_host == NULL)
+  {
+    g_chrony_host = strdup(CHRONY_DEFAULT_HOST);
+    assert(g_chrony_host);
```
Please handle the error, instead of asserting.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+    ERROR(PLUGIN_NAME ": Error receiving packet: %s (%d)", strerror(errno),
+          errno);
+    return CHRONY_RC_FAIL;
+  } else
```
put the else on the next line.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+    if (chrony_connect() == 0)
+    {
+      g_is_connected = 1;
+    } else
```
put the else on the next line.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+    }
+
+    if (valid_command == 0)
+    {
```
we tend not to use braces for single-line if statements.

bbczeuz replied:
Is this negotiable? I once spent more than a day fixing some strange error in some other code that only happened due to missing braces, since then I consequently use braces everywhere.

rubenk replied:
We've had a similar discussion already. I personally don't disagree with your preference for bracing, but you want your code included in an existing project with an existing style, so you should adhere to it. If each plugin used a different style the codebase would quickly become a mess.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+          seq_nr);
+    if (chrony_send_request(p_req, req_size) != 0)
+    {
+      break;
```
same here

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+    DEBUG(PLUGIN_NAME ": Waiting for response");
+    if (chrony_recv_response(p_resp, resp_size, p_resp_size) != 0)
+    {
+      break;
```
and here.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+  coef = uval % (1U << FLOAT_COEF_BITS);
+  if (coef >= 1 << (FLOAT_COEF_BITS - 1))
+  {
+    coef -= 1 << FLOAT_COEF_BITS;
```
and here.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+  exp = (uval >> FLOAT_COEF_BITS) - FLOAT_COEF_BITS;
+  if (exp >= 1 << (FLOAT_EXP_BITS - 1))
+  {
+    exp -= 1 << FLOAT_EXP_BITS;
```
and here.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+      ERROR(PLUGIN_NAME ": Error duplicating host name");
+      return CHRONY_RC_FAIL;
+    }
+  } else if (strcasecmp(p_key, CONFIG_KEY_PORT) == 0)
```
put the else on a new line please.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+      ERROR(PLUGIN_NAME ": Error duplicating port name");
+      return CHRONY_RC_FAIL;
+    }
+  } else if (strcasecmp(p_key, CONFIG_KEY_TIMEOUT) == 0)
```
same here.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+  {
+    time_t tosec = strtol(p_value, NULL, 0);
+    g_chrony_timeout = tosec;
+  } else
```
and here.

rubenk commented on src/chrony.c:
```diff
@@ -0,0 +1,1147 @@
+  }
+  if (g_chrony_host != NULL)
+  {
+    free(g_chrony_host);
```
please use sfree() which sets the pointer to NULL too.

rubenk commented:
Thanks @bbczeuz!

eyalzek commented:
Hey, is there a way to use this plugin currently? I'm on collectd 5.5.2.git and trying to load this plugin I'm getting:
`plugin_load: Could not find plugin "chrony" in /usr/lib/collectd`
